File: baekjoon/python/gold5/q21610_magician_shark_and_rain_hoper.py
#https://www.acmicpc.net/problem/21610

import sys
input = sys.stdin.readline

def solution():
    moves = [(), (0, -1), (-1, -1), (-1, 0), (-1, 1), (0, 1), (1, 1), (1, 0), (1, -1)]
    n, m = map(int, input().split())
    board = []
    for _ in range(n):
        board.append(list(map(int, input().split())))
    clouds = [(n-2, 0), (n-2, 1), (n-1, 0), (n-1, 1)]
    for _ in range(m):
        d, s = map(int, input().split())
        movex = moves[d][0]
        movey = moves[d][1]
        disappeared_clouds = []
        for cloud in clouds:
            nx = cloud[0] + (movex * s)
            ny = cloud[1] + (movey * s)
            if nx < 0 :
                nx = n - ((-nx) % n)
                if nx == n :
                    nx = 0
            if nx >= n :
                nx = nx % n
            if ny < 0 :
                ny = n - ((-ny) % n)
                if ny == n :
                    ny = 0
            if ny >= n :
                ny = (ny) % n 
            board[nx][ny] += 1
            disappeared_clouds.append((nx, ny))
        clouds = []
        for cloud in disappeared_clouds :
            x = cloud[0]
            y = cloud[1]
            cnt = 0
            for move in [moves[2], moves[4], moves[6], moves[8]] :
                nx = x + move[0]
                ny = y + move[1]
                if 0 <= nx < n and 0 <= ny < n :
                    if board[nx][ny] >= 1:
                        cnt += 1
            board[x][y] += cnt
        for i in range(n):
            for j in range(n):
                if (i, j) in disappeared_clouds : 
                    continue
                if board[i][j] >= 2 :
                    board[i][j] -= 2
                    clouds.append((i, j))
    result = 0
    for i in board :
        result += sum(i)
    return result

# print(solution())





    
            

File: baekjoon/java/src/gold5/Q21610_MagicianSharkAndRainHoper.java
package gold5;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

// https://www.acmicpc.net/problem/21610
public class Q21610_MagicianSharkAndRainHoper {
    private static int[][] board;
    private static int[][] clouds;
    private static int[][] moves = new int[][]{{0, 0}, {0, -1}, {-1, -1}, {-1, 0}, {-1, 1}, {0, 1}, {1, 1}, {1, 0}, {1, -1}};

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        int n = Integer.parseInt(st.nextToken());
        int m = Integer.parseInt(st.nextToken());
        board = new int[n][n];
        clouds = new int[n][n];

        for (int i = 0; i < n; i++) {
            st = new StringTokenizer(br.readLine());
            for (int j = 0; j < n; j++) {
                board[i][j] = Integer.parseInt(st.nextToken());
            }
        }

        clouds[n - 1][0] = 1;
        clouds[n - 1][1] = 1;
        clouds[n - 2][0] = 1;
        clouds[n - 2][1] = 1;

        for (int i = 0; i < m; i++) {
            st = new StringTokenizer(br.readLine());
            int d = Integer.parseInt(st.nextToken());
            int s = Integer.parseInt(st.nextToken());
            int moveX = moves[d][0] * s + s * n;
            int moveY = moves[d][1] * s + s * n;
            int[][] disappearedClouds = new int[n][n];
            for (int j = 0; j < n; j++) {
                for (int k = 0; k < n; k++) {
                    if (clouds[j][k] == 1) {
                        int nx = (j + moveX) % n;
                        int ny = (k + moveY) % n;
                        disappearedClouds[nx][ny] = 1;
                        board[nx][ny] += 1;
                    }
                }
            }

            for (int j = 0; j < n; j++) {
                for (int k = 0; k < n; k++) {
                    if (disappearedClouds[j][k] == 1) {
                        for (int[] move : new int[][]{moves[2], moves[4], moves[6], moves[8]}) {
                            int nx = j + move[0];
                            int ny = k + move[1];
                            if (0 <= nx && nx < n && 0 <= ny && ny < n) {
                                if (board[nx][ny] >= 1) {
                                    board[j][k] += 1;
                                }
                            }
                        }
                    }
                }
            }

            for (int j = 0; j < n; j++) {
                for (int k = 0; k < n; k++) {
                    if (disappearedClouds[j][k] == 1) {
                        disappearedClouds[j][k] = 0;
                        continue;
                    }
                    if (board[j][k] >= 2) {
                        board[j][k] -= 2;
                        disappearedClouds[j][k] = 1;
                    }
                }
            }
            clouds = disappearedClouds;
        }

        int sum = 0;
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                sum += board[i][j];
            }
        }
        System.out.println(sum);
    }
}
